odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+
+
+class FunctionalSum(Functional, OperatorSum):
+
+    """Expression type for the sum of functionals.
+
+    ``FunctionalSum(func1, func2) <==> (x --> func1(x) + func2(x))``.
+    """
+
+    def __init__(self, func1, func2, tmp_dom=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func1, func2 : `Functional`
+            The summands of the functional sum. Their `Operator.domain`
+            and `Operator.range` must coincide.
+        tmp_dom : `Operator.domain` `element`, optional
+            Used to avoid the creation of a temporary when applying the
+            gradient.
+        """
+        # TODO: Tmp_dom, what is it used for?
+        if not isinstance(func1, Functional):
+            raise TypeError('functional 1 {!r} is not a Functional instance.'
+                            ''.format(func1))
+        if not isinstance(func2, Functional):
+            raise TypeError('functional 2 {!r} is not a Functional instance.'
+                            ''.format(func2))
+
+        Functional.__init__(self, domain=func1.domain,
+                            linear=(func1.is_linear and func2.is_linear),
+                            grad_lipschitz=(func1.grad_lipschitz +
+                                            func2.grad_lipschitz))
+
+        OperatorSum.__init__(self, func1, func2, tmp_ran=None,
+                             tmp_dom=tmp_dom)
+
+    @property
+    def gradient(self):
+        """Gradient operator of functional sum."""
+        return self.left.gradient + self.right.gradient
+
+    def derivative(self, point):
+        """Returns the derivative operator in the given point.
+
+        This function returns the linear operator
+
+            ``x --> <x, grad_f(point)>``,
+
+        where ``grad_f(point)`` is the gradient of the functional in the point
+        ``point``.
+
+        Parameters
+        ----------
+        point : `LinearSpaceVector`
+            The point in which the gradient is evaluated.
+
+        Returns
+        -------
+        out : `DerivativeOperator`
+            The linear operator that maps ``x --> <x, grad_f(point)>``.
+        """
+        return self._op1.derivative(point) + self._op2.derivative(point)
+
+
+class FunctionalScalarSum(Functional, OperatorSum):
+
+    """Expression type for the sum of a functional and a scalar.
+
+    ``FunctionalScalarSum(func, scalar) <==> (x --> func(x) + scalar)``
+    """
+
+    def __init__(self, func, scalar, tmp_dom=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            Functional to which the scalar is added.
+        scalar : `element` in the `field` of the ``domain``
+            The scalar to be added to the functional. The `field` of the
+            ``domain`` is the range of the functional.
+        tmp_dom : `LinearSpaceVector`, optional
+            ...
+        """
+        # TODO: what is tmp_dom used for?
```

Comment:
Temporary for the result of `ConstantOperator` here, not needed. Instead you could create `constant * one()` by hand, then there really is no need for a temporary.